### PR TITLE
Load data layer handlers dynamically

### PR DESCRIPTION
**Pull request by dmsnell**
## TL;DR

Instead of building all of the data layer handlers into the initial bundle have webpack import them as-needed at the first time it's possible that an action they are listening for could be dispatched.

 - shrink initial bundle
 - speed up app boot
 - free derivative works of Calypso SDK from including all the handlers (start _tabula rasa_ when using the data layer)

## Details

Currently the data layer handlers have a static relationship with the app. Since they must all load at boot they all end up in the main `build` bundle. This bloats the build and also means that all handlers must initialize at app boot which slows down the process.

In this PR I'm working on separating the static initialization into dynamically-loaded handlers that asynchronously load in the bundles where they are needed.

The primary means of operation is designed around the primary challenge of making sure that the right handlers are in place at the time any actions they watch are dispatched. Long-term this applies to reducers as well and I believe that the system here can be used with reducers; I believe that we can eventually automate this and remove the manual work.

The action creators that dispatch the actions which the handlers listen to are currently split among separate bundles. If we have a dynamic loading system for handlers then all we theoretically need to ensure is that the handlers are at least in the same bundle as the first action creator sending those actions. I have created a new function called `requireHandlers()` which provides that guarantee. This couples with one change to the handler files: requiring the appropriate handlers in itself when loaded; and one change to the action creator files: importing the handler file when loaded.

```js
// state/data-layer/some/handler/index.js
registerHandlers( [
  'state/data-layer/some/handler',
  { PLANS_REQUEST: [ dispatchRequestEq( { fetch } ) ] }
];

// some/action/creator/actions.js
import { PLANS_REQUEST } from 'state/action-types';

// note that we aren't importing any bindings, just initializing the module
import 'state/data-layer/some/handler';

export const doSomething = ( { … } );
```

By adding this idempotent and manual constraint on the handler loading in the action creator we should never have a mismatch between dispatches and handlers that matters.

## Testing

You can verify that the plans handlers only load when visiting the plans page or when something depends on the plans information.

Start at the stats screen and issue this command into the console dispatcher.

```js
dispatch( { type: 'PLANS_REQUEST' } );
```

The action should pass through with no corresponding network activity.

Visit the plans page. In my test the actual handlers were incorporated into the `jetpack~connect~plans~signup.js` bundle. After loading, repeat the dispatch from above. This time you should see a network request issued to `/sites/${ siteId }/plans` as a side-effect of the dispatch. This confirms that the handler is in place.

Verify that we don't have places that call for plans information but where we aren't loading them.

## Size Diff
http://iscalypsofastyet.com/branch?branch=framework/async-data-layer
```
Delta:
chunk                                        stat_size           parsed_size           gzip_size
jetpack-connect~plans~signup                   +2826 B  (+3.5%)       +680 B  (+1.6%)     +247 B  (+2.3%)
zoninator                                         +5 B  (+0.0%)        -15 B  (-0.0%)       -9 B  (-0.1%)
wp-job-manager                                    +5 B  (+0.0%)        -15 B  (-0.0%)      -10 B  (-0.1%)
woocommerce                                       +5 B  (+0.0%)        -15 B  (-0.0%)       -5 B  (-0.0%)
checkout                                       +2826 B  (+0.7%)       +680 B  (+0.3%)     +217 B  (+0.5%)
async-load-components-web-preview-component    +2826 B  (+2.7%)       +680 B  (+1.2%)     +269 B  (+2.1%)
manifest                                          +0 B                  +0 B                -6 B  (-0.1%)
async-load-design-blocks                       +2826 B  (+0.2%)       +680 B  (+0.1%)     +204 B  (+0.1%)
build                                          -4204 B  (-0.1%)       -501 B  (-0.0%)     +254 B  (+0.1%)
```

### CODE REVIEW

File: client/state/data-layer/extensions-middleware.js [hunk 1]
```diff
@@ -68,6 +68,8 @@ export default store => next => {
 	configuration.store = store;
 	configuration.next = next;
 
+	return action => next( action );
```
Comment (dmsnell) [inline client/state/data-layer/extensions-middleware.js]: ignore this - i was trying to work around something temporarily

File: client/state/data-layer/wpcom-api-middleware.js [hunk 2]
```diff
@@ -46,7 +57,7 @@ const shouldNext = action => {
  * WordPress.com API and passes them off to the
  * appropriate handler.
  *
- * @see state/utils/local indicates that action should bypass data layer
+ * @see moduleState/utils/local indicates that action should bypass data layer
```
Comment (dmsnell) [inline client/state/data-layer/wpcom-api-middleware.js]: junk

**Comment by coderkevin** [comment]
I'm afraid I'm not quite understanding what this PR is trying to accomplish. 🤔

File: client/state/activity-log/actions.js [hunk 3]
```diff
@@ -29,6 +31,8 @@ import {
 	REWIND_BACKUP_DISMISS_PROGRESS,
 } from 'state/action-types';
 
+requireHandlers( handlers );
```
Comment (coderkevin) [inline client/state/activity-log/actions.js]: You know, this could be called from the action creator, which would be an even lazier load. It wouldn't introduce the handlers until just before the action is dispatched.

Comment (dmsnell) [reply client/state/activity-log/actions.js]: the worry with that is the asynchronous loading nature. if we only lazy-load it at the action creator then we are _guaranteed_ to dispatch the action before the handler is loaded (event loop processing)

on the other hand it does give us reason to split out action creators when there is a module with lots of them but only a couple which require a specific handler set

File: client/state/data-layer/handler-loading.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+		}
+
+		requiredHandlers.add( id );
+		nextHandlers = mergeHandlers( nextHandlers, handlers );
+	} );
+
+	if ( state.handlers !== nextHandlers ) {
+		state.handlers = nextHandlers;
+	}
+};
```
Comment (dmsnell) [inline client/state/data-layer/handler-loading.js]: @coderkevin this looks a lot like what you are doing in the `extensions` directory. I'm not wanting to insult you by repeating it here, just wasn't sure how to pull in the code you wrote.

**Comment by jsnajdr** [comment]
This looks like a nice improvement :+1:

I'm only very confused by how things are named. Instead of `requireHandlers` function and `requiredHandlers` map, wouldn't `registerHandlers` make more sense? That's what the function is doing, isn't it? Registering handlers in a registry.

About the question from where we should register handlers:
I think that a very good place to register many handlers is the extension's/section's `index.js` file. For example, `client/my-sites/plans/index.js` registers `page` router handlers for the Plans section. If Plans have their own reducer and effects handlers, this looks like one of good places to register them, doesn't it?

There are two similar, but different ways how a section can ask for handlers being registered:
1. I'm a Plans section and I want to work with "Products List" data module that's part of framework. Calypso, please make sure that it's loaded and available for me.
2. I'm a Plans section and I want to register my private "Plans" data module. The code is part of my bundle. Calypso, here is the code, please register it so that I can use it.

Do we declare both kinds of dependencies the same way? No 1 "requires" them, no 2 "registers" them.

**Comment by blowery** [comment]
As I was reading this, I was expecting to come across an async import at some point, but never did. After further reflection, I realized that what's happening here is a rearrangement of the dependency graph, which results in the data handler code being loaded later. I like that, as it means we don't have to worry as much about what happens it the chunk containing the handler fails to load. It's already covered by our existing mechanism.

I agree that we could reuse something like this for the reducer tree. In that case, we'd likely have to add imports from the selectors and action creators? Reducers also have the complication that dispatch is expected to be sync, while we'd be making it effectively async when the reducer hasn't yet been loaded... Something I think we can push down the road for now though.

**Comment by blowery** [comment]
+1 on calling it `registerHandlers` rather than `requireHandlers`, as there's no `require` happening and I think `register` maps better to what's happening.

**Comment by dmsnell** [comment]
> Reducers also have the complication that dispatch is expected to be sync, while we'd be making it effectively async when the reducer hasn't yet been loaded... Something I think we can push down the road for now though.

Well it's not async because by the first time the action _could_ be dispatched webpack will have already included the new reducers in the code in the browser so it can all remain synchronous. by using webpack's code-splitting we can eliminate this concern

> +1 on calling it registerHandlers rather than requireHandlers, as there's no require happening and I think register maps better to what's happening.

fine. I don't like that name as I feel like it communicates something else. `require` to me means "I need to make sure this is here" whereas `register` means "add this thing in"

granted, it started differently, so `register` probably as you said _does_ make more sense given the code

**Comment by blowery** [comment]
> Well it's not async because by the first time the action could be dispatched webpack will have already included the new reducers in the code in the browser so it can all remain synchronous. by using webpack's code-splitting we can eliminate this concern

Ah, yes, quite right! I still had my original expectations in my head when I wrote that.

**Comment by dmsnell** [comment]
of course…removing 1500 bytes from the source code in `build` led to a 500 byte growth in the gzip output 🙃 

I expect that `build` will shrink as we pull out more handlers

File: client/state/data-layer/handler-loading.js (hunk 4 above)
Comment (coderkevin) [reply client/state/data-layer/handler-loading.js]: Would this replace the code there? If so, I'm cool with it. I just don't want it in two places if at all possible.

Comment (dmsnell) [reply client/state/data-layer/handler-loading.js]: not unless you wanted. on that note:

 - do we deregister any reducers/handlers now?
 - if so, I think we could simplify this further with a `Map()` instead of a set. would you have interest in my tossing out an idea?

**Comment by dmsnell** [comment]
> I think that a very good place to register many handlers is the extension's/section's index.js file. For example, client/my-sites/plans/index.js registers page router handlers for the Plans section. If Plans have their own reducer and effects handlers, this looks like one of good places to register them, doesn't it?

that's a good question @jsnajdr but the reason I have them in the action files is to match dependencies. the extensions/sections don't actually need them, it's the point where the action _could_ be dispatched that needs them. that means if we pair the handlers with the action creators we will guarantee the maximum code-splitting efficiency, I think.

ultimately I want to do this automatically. consider this idea:
 - scan all files in `state/**/actions.js`
 - build map of which ones import which action types
 - scan all files in `state/data-layer`
 - build map of which ones import which action types
 - for each action type in the second set, make sure that the registration is in the first set

hope that makes sense

File: client/state/data-layer/handler-loading.js (hunk 4 above)
Comment (dmsnell) [reply client/state/data-layer/handler-loading.js]: maybe we can consider ripping out the de-registration?

**Comment by blowery** [comment]
> that's a good question @jsnajdr but the reason I have them in the action files is to match dependencies. the extensions/sections don't actually need them, it's the point where the action could be dispatched that needs them.

I also like having the registration close to the point of usage. If we had it up at the section definition, it gets really easy for that mapping to go stale, since now the person making changes to an action creator has to know to also go poking around in section definitions.

File: client/state/data-layer/handler-loading.js (hunk 4 above)
Comment (coderkevin) [reply client/state/data-layer/handler-loading.js]: I'm okay with that as long as we're sure we won't need to ever deregister anything.

**Review by blowery — COMMENTED** [review]
LGTM! Looking forward to moving all of the data layer this way.

**Review by blowery — APPROVED** [review]

File: client/state/data-layer/handler-loading.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
+/** @format */
+/**
+ * Internal dependencies
+ */
+import { mergeHandlers } from 'state/action-watchers/utils';
+
+export const state = {
+	handlers: {},
+};
```
Comment (jsnajdr) [inline client/state/data-layer/handler-loading.js]: What if the `state` variable was called `registeredHandlers` instead? Right now, I have to think very hard to understand the code: is `state` the Redux state object, like everywhere else, or something different? The name doesn't help me at all.

Also, what if the `registeredHandlers` was a mutable object without the `handlers` subproperty, and `registerHandlers` just merged new properties into it? The code that updates the registered handlers seems unnecessarily complicated. For example, why is there the `if ( state.handlers !== nextHandlers )` check? Does it prevent anything bad from happening?

Of course, an ideal API for this module would be to export just two functions: `registerHandlers` and `findHandlers` and keep the actual data structure private to the module. Then we can modify the structure however we want (possibly to allow unregistration) without changing the interface.

These are just optional, although very desirable, improvements. The PR will work very well even without them.

**Comment by jsnajdr** [comment]
> ultimately I want to do this automatically. consider this idea

Thanks for the explanation @dmsnell, it makes sense. By "automatically", you mean something like a codemod? Or something more sophisticated like a really clever Babel transform or Webpack plugin? :)

File: client/state/data-layer/handler-loading.js (hunk 5 above)
Comment (dmsnell) [reply client/state/data-layer/handler-loading.js]: excellent suggestions @jsnajdr - in fact I think these are evidence of the experimental nature of how I wrote this. I like your feedback so well I'm going to go ahead and make these changes. internalizing the abstraction is A+++

it wasn't until I discovered that a circular dependency was behind a nasty bug that I split out the registration function into a separate module. artifacts all over…

**Comment by dmsnell** [comment]
> By "automatically", you mean something like a codemod? Or something more sophisticated like a really clever Babel transform or Webpack plugin? :)

Either really. I'd rather test with a codemod first that we can apply and then consider a webpack plugin. The tradeoffs in my mind are certainty vs. build speed.

**Comment by dmsnell** [comment]
@jsnajdr updated here with your suggestions

@coderkevin because the end result seemed much simpler I removed the extensions middleware. seems like this should be fine - I couldn't find any use of the `config` parameter to `addHandlers` which made for a quick port. if we need to remove handlers in the future I think that will be a trivial addition, but no need now

**Review by jsnajdr — APPROVED** [review]
Thanks @dmsnell for improving the module API.

I have one more little suggestion: the `registerHandlers` function will be mostly called only with one handler object at once. What if we supported the following syntax and saved a considerable number of square brackets?
```js
registerHandlers( 'state/data-layer/wpcom/foo', {
  [ FOO_REQUEST ]: fooDispatcher,
  [ GOO_REQUEST ]: gooDispatcher,
} );
```
The variant with array can still work, but it's a special case.

Even if you don't implement the suggestion, this is certainly ready to :ship:. Looking forward to see the ICFY charts go down again!

**Comment by blowery** [comment]
@dmsnell login appears to be broken:
1) go to http://calypso.localhost:3000/log-in
2) enter username, hit continue
3) every grays out, nothing happens

File: client/state/data-layer/wpcom-api-middleware.js [hunk 6]
```diff
@@ -3,21 +3,20 @@
...
 import thirdPartyHandlers from './third-party';
 import wpcomHandlers from './wpcom';
 
-const mergedHandlers = mergeHandlers(
-	httpData,
-	httpHandlers,
-	wpcomHttpHandlers,
-	thirdPartyHandlers,
-	wpcomHandlers
+registerHandlers(
```
Comment (blowery) [inline client/state/data-layer/wpcom-api-middleware.js]: this call doesn't match. `registerHandlers` expects `( id, handlers )`

Comment (dmsnell) [reply client/state/data-layer/wpcom-api-middleware.js]: derp. fixed.
